src/test/java/io/confluent/examples/streams/ReorderIntegrationTest.java

```diff
@@ -0,0 +1,239 @@
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * End-to-end integration test that demonstrates how to reorder the stream of incoming messages
+ * by the timestamp embedded in the message payload.
+ * <p>
+ * Makes sense only on per partition basis.
+ * <p>
+ * Reordering occurs within the time windows defined by the {@code grace} constructor parameter.
+ *
+ * Note: This example uses lambda expressions and thus works with Java 8+ only.
+ */
+public class ReorderIntegrationTest {
+
+  public static class ReorderTransformer<K, V>
+      implements Transformer<K, V, KeyValue<K, V>> {
+
+    public interface StoreKeyGenerator<K, V> {
+      K getStoreKey(K key, V val);
+    }
+
+    private final String storeName;
+    private final Duration grace;
+    private KeyValueStore<K, V> reorderStore;
+    private ProcessorContext context;
+    private final StoreKeyGenerator<K,V> storeKeyGenerator;
+
+    public ReorderTransformer(final String storeName, final Duration grace,
+                              final StoreKeyGenerator<K,V> storeKeyGenerator) {
+      this.storeName = storeName;
+      this.grace = grace;
+      this.storeKeyGenerator = storeKeyGenerator;
+    }
+
+    @Override
+    public void init(final ProcessorContext context) {
+      this.reorderStore = context.getStateStore(this.storeName);
+      this.context = context;
+      context.schedule(
+          this.grace,
+          PunctuationType.STREAM_TIME,
+          this::punctuate
+      );
+    }
+
+    /**
+     * 1) read the timestamp from the message value
+     * 2) inserts into a KeyValueStore using (timestamp, message-key) pair
+     * as the key and the message-value as the value, this also provides
+     * de-duplication.
+     *
+     * @param key
+     * @param value
+     * @return
+     */
+    @Override
+    public KeyValue<K, V> transform(final K key, final V value) {
+      // Keys need to contain and be sortable by time
+      final K storeKey = storeKeyGenerator.getStoreKey(key, value);
+      final V storeValue = reorderStore.get(storeKey);
+      if(storeValue == null) {
+        reorderStore.put(storeKey, value);
+      }
+      return null; 	// null suppresses sending to downstream
+    }
+
+    /**
+     * Scheduled to be called automatically when the period
+     * within which message reordering occurs expires.
+     *
+     * Outputs downstream accumulated records sorted by their timestamp.
+     *
+     * 1) read the entire store
+     * 2) send the fetched messages in order using context.forward() and delete
+     * them from the store
+     *
+     * @param timestamp – stream time of the punctuate function call
+     */
+    void punctuate(final long timestamp) {
+      try(KeyValueIterator<K, V> it = reorderStore.all()) {
+        while (it.hasNext()) {
+          final KeyValue<K, V> kv = it.next();
+          context.forward(kv.key, kv.value);
```

Comment:
I think this implementation is flawed. Ordering only makes sense per partition/key. 

Say incoming records have key `k1`, and there are two messages with timestamps `t0`, and `t1`. The `StoreKeyGenerator` will map that into some combination of the original key and the record timestamp, say something simple like `k1_t0` and `k1_t1`. However, when you forward these message with the "composite" store key, they'll likely end up in two different partitions. Hence, although you sorted them, any downstream processor/consumer has no guarantee in the order after the re-partition. The correct thing to do would be to preserve the original key, and use that to forward the message.

Reply:
It is an example. It can be extended with either additional store for composite to original key mapping or a second interface that would allow extracting original key from the composite one before passing downstream.